test/unit/specs/components/DataCollector/index.spec.js

```diff
@@ -101,11 +99,11 @@ describe("Event Command", () => {
     const deferred = defer();
     onBeforeEventSpy.and.returnValue(deferred.promise);
     eventCommand({});
-    return flushPromises().then(() => {
+    return flushPromiseChains().then(() => {
       expect(lifecycle.onBeforeEvent).toHaveBeenCalled();
       expect(network.sendRequest).not.toHaveBeenCalled();
       deferred.resolve();
-      flushPromises().then(() => {
+      flushPromiseChains().then(() => {
         expect(network.sendRequest).toHaveBeenCalled();
       });
     });
```

Comment:
There's a different way to do this with `expectsAsync` instead of `flushPromises`, but I feel like it might be more difficult to understand:

```
it("Allows other components to hold up the event", () => {
    const deferred = defer();
    const onBeforeEventCalledPromise = new Promise(resolve => {
      onBeforeEventSpy.and.callFake(() => {
        resolve();
        return deferred.promise;
      });
    });
    const sendRequestCalledPromise = new Promise(resolve => {
      network.sendRequest.and.callFake(() => {
        resolve();
        return Promise.resolve({});
      });
    });
    eventCommand({});
    return expectAsync(onBeforeEventCalledPromise)
      .toBeResolved()
      .then(() => {
        expect(network.sendRequest).not.toHaveBeenCalled();
        deferred.resolve();
        return expectAsync(sendRequestCalledPromise).toBeResolved();
      });
  });
```

Reply:
Interesting. I guess it would be cool to use built-in APIs instead of our utils, but I'll leave it to you.

Reply:
Well now that you put it that way..........
I'm totally on the fence about it.

Reply:
I'm not sure that this proposed test would actually test the functionality.  I think it would depend on the exact details of the promise implementation.  (and maybe the number of promises chained together in the implementation?) . I think running through 10 promises does a better job at making sure all the promises are flushed.

Reply:
I think I understand it a little better now.  I think the second flushPromises could be replaced with an expectAsync, but not the first.

Reply:
So you can use expectAsync to verify that something happens, but I don't think you can use it to verify that something didn't happen.

Reply:
That's correct. In both usages of `expectAsync` here, we are verifying that something happened. `expectAsync(onBeforeEventCalledPromise).toBeResolved()`, waits until `onBeforeEvent` is called. At that point in time, `expect(network.sendRequest).not.toHaveBeenCalled();` checks to see that `sendRequest` has not been called. We then resolve the promise returned from `onBeforeEvent`. At that point in time, `expectAsync(sendRequestCalledPromise).toBeResolved()` then waits until `network.sendRequest` is called.

The weird part to me is that we're using `expectAsync` as sort of a control flow for our test code and not actually an assertion that a promise returned by our _source_ code is being resolved/rejected as expected.

Reply:
I think from my last comment and sensing some confusion on your part would indicate that using `flushPromiseChains` is preferable.

Reply:
Yes, I think replacing the first `expectAsync(onBeforeEventCalledPromise).then(...)` with `onBeforeEventCalledPromise.then(...)` would mean the same thing, so I think you need to let the promises flush to be sure that sendRequest wasn't going to be called anyways.

Reply:
Ah, I see what you're saying. Yeah, I think it's probably safer to use `flushPromiseChains`. Thanks for the feedback, @jonsnyder.